### ISSUE TITLE
Names and formats for mismatch_rate and recombination_rate parameters

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/tsinfer#303]
(Continuing on the discussion from #273 ), currently if we provide either the `mismatch_rate` or the `recombination_rate` parameter to match_samples or match_ancestors, it is expected to be an array of length the number of inference sites. These give the switching vs emission probabilities in the LS HMM. One thing to bear in mind is that with https://github.com/tskit-dev/tsinfer/pull/266 it may not be obvious how long to make this array. But it might also be that we want at least the recombination_rate to be specified as a true *rate* per unit sequence, akin to the recombination_rate parameter of msprime, rather than a set of values which will vary depending on the distance between adjacent sites (which is what we do currently, and which I think of as more of a probability than a rate).

For recombination, It would certainly make sense to me to have the `recombination_rate` parameter set to either a constant per-base rate for the whole genome, or possibly an msprime.RecombinationMap. That way we don't need to worry about the inference sites thing. For experts, we could provide an alternative parameter, say `recombination_distances` or `sitewise_recombination_probabilities` or something, that explicitly sets the switch probabilities for each site.

I don't have any particularly well formed ideas for the mismatch rate. I tend to set each value in the mismatch array to a constant for the whole genome, based on the median recombination probability, rather than setting each value as a proportion of the recombination probability at the previous site. I'm not sure what a sensible name or expected format for this parameter should be, so would welcome some discussion.

**Comment by hyanwong** [comment]
This should be for 0.2.0, and should use the intervals.py file that I am reworking in https://github.com/tskit-dev/msprime/pull/1185 (see also https://github.com/tskit-dev/msprime/issues/1255#issuecomment-725535906)

**Comment by hyanwong** [comment]
OK, so here's what I imagine doing. We include the new intervals.py file from msprime into tsinfer, to give us a `tsinfer.RateMap` class. For `tsinfer.infer` we have a new `recombination_rate` parameter which we just pass down to `match_samples` and `match_ancestors`, but we don't have a `mismatch_rate` parameter - if you want to set that to something other than the default of 1, you need to run MA and MS explicitly.

The match_ancestors `recombination_rate` parameter, could be documented something like this (with a similar wording for match_samples):

```
:param recombination_rate: Either a RateMap object, or a floating point value
    giving a single rate :math:`rho` across the entire sequence.
    Integrating this rate between adjacent sites provides a list of relative
    likelihoods of recombination, :math:`r`, which determine the trade off between 
    recombination and multiple mutation during the matching process. More specifically,
    if a conflicting site is found while matching an ancestor against older haplotypes,
    then the higher the recombination likelihood compared to the site's mismatch
    likelihood, the more likely the conflict at that site will be resolved by
    postulating an additional recombination event as opposed to an additional mutation.
```

For `mismatch_rate`, I am toying with the idea that we should, by default, allow a single value which is *relative* to the median recombination rate (and defaults to 1), but if a map is given, it specifies *absolute* values. The suggested documentation is below - do you think it is too confusing to switch between relative and absolute rates like this @jeromekelleher , or does the documentation make the distinction clear enough? I think a single relative value is more immediately intuitive, but it's undesirable to use relative values in a RateMap, as the meaning will change for different datasets (e.g. different spacing of site positions).

```
:param mismatch_rate: Either a floating point value or a RateMap object. 
    * If a single value, this specifies the *relative* mismatch
      rate for all sites, compared to the recombination likelihoods. More specifically
      the list of recombination likelihoods :math:`r` between adjacent sites are
      calculated, and the mismatch likelihood is given by the median value
      :math:`$\tilde{r}` multiplied by this value. This relative mismatch rate is
      more intuitive than an absolute value, as the balance between
      recombination and recurrent mutations is determined by the ratio of the
      recombination to mismatch likelihoods. Relative rates somewhere between 1e-3 and
      1e3 are generally recommended.
    * If a RateMap object, this is interpreted as specifying *absolute* mismatch rates:
      this is, the mismatch likelihood at a site is simply the rate value
      at that position. This is a more flexible way to specify mismatch, and can be used,
      for example, to account for variable mutation rates at different sites. Note,
      however, that the magnitude of the rates in the map should therefore
      be compared to the magnitude of the average recombination rate multiplied by the
      average distance between sites, and the values used will be specific to the
      dataset being analysed.
```

**Comment by awohns** [comment]
This looks good to me @hyanwong . I would say that the descriptions could be shortened. I've included three bits which feel more like justification/explanation of the algorithm than descriptors of the parameters:  
```
More specifically, if a conflicting site is found while matching an ancestor against older haplotypes,
    then the higher the recombination likelihood compared to the site's mismatch
    likelihood, the more likely the conflict at that site will be resolved by
    postulating an additional recombination event as opposed to an additional mutation.
```
```
Specifying a relative mismatch rate is more intuitive than using absolute rates,
      as the balance between recombination and recurrent mutations is determined by the ratio of the
      recombination to mismatch likelihoods. Relative rates somewhere between 1e-3 and
      1e3 are generally recommended.
```
```
Note,  however, that the magnitude of the rates in the map should therefore
      be compared to the magnitude of the average recombination rate multiplied by the
      average distance between sites, and the values used will be specific to the
      dataset being analysed.
```

**Comment by hyanwong** [comment]
Would you find it confusing if, as suggested, the mismatch rate is relative when specified as a single number, but absolute when specified as a RateMap, @awohns? That's the main thing that I'm not sure about.

**Comment by awohns** [comment]
I think it's alright, as it avoids the `np.median(recombination_rates)` business if the users wants a specify a single number

**Comment by hyanwong** [comment]
The other question is whether we want the ability - that we do have at present - to explicitly set the likelihood values, using an array of the same length as the number of inference sites (for mismatch) or an array one item shorter (for recombination). If we do want to be able to set these directly, we could either 

1. overload the `recombination_rate` or `mismatch_rate` parameters, such that if an array is passed (rather than a single number or a RateMap) then it specifies the likelihoods directly, or
2. have a separate, possibly undocumented, pair of parameters, perhaps `recombination_likelihoods` and `mismatch_likelihoods`, that can be used. Then we would need to check that only one of `recombination_rate` and `recombination_likelihoods` is set, etc.

I'm not sure what is preferable here.

**Comment by hyanwong** [comment]
Here's roughly what we decided upon in conversation @jeromekelleher. At some point do you want to glance over it and check that all seems reasonable, then I can code it up. I don't know if it is all too wordy for documentation outside of a tutorial, but I've done my best to make it concise. Could you also take a look @awohns , given your helpful comments previously? I'm assuming the `match_samples` documentation will look much the same.

(note that I assume recombination_rate can be a RateMap, which is the format returned by the new `read_hapmap` code, but we might want to subclass it into a more specific `RecombinationMap` class - this is all detail to be sorted out in https://github.com/tskit-dev/msprime/pull/1185 )

```
def match_ancestors(
    sample_data,
    ancestor_data,
    *,
    recombination_rate=None,
    relative_mismatch=None,
    path_compression=True,
    num_threads=0,
    precision=None,
    mismatch_likelihoods=None,    
    # The following are deliberately undocumented
    recombination_likelihoods=None,  # analogous to `mismatch_likelihoods` for recombination
    extended_checks=False,
    engine=constants.C_ENGINE,
    progress_monitor=None,
):
    """
    match_ancestors(
        sample_data,
        ancestor_data,
        *,
        recombination_rate=None,
        relative_mismatch=None,
        path_compression=True,
        num_threads=0,
        precision=None,
        mismatch_likelihoods=None)

    Runs the ancestor matching :ref:`algorithm <sec_inference_match_ancestors>`
    on the specified :class:`SampleData` and :class:`AncestorData` instances,
    returning the resulting :class:`tskit.TreeSequence` representing the
    complete ancestry of the putative ancestors.

    The recombination and mismatch parameters set a likelihood of mismatch at a given
    site and a likelihood of recombination between this site and the previous one.
    These ratio of these likelihoods determines what happens if the site is a cause of
    conflict when matching an ancestor against older haplotypes. In this case, the
    higher the ratio of mismatch to recombination likelihoods,  then the greater the
    odds the conflict will be resolved by postulating an additional mutation as opposed
    to an additional recombination event. Setting `relative_mismatch` to very high values
    therefore infers tree sequences with many recurrent mutations and relatively fewer
    recombinations (and edges). Conversely, a low mismatch to recombination likelihood
    ratio will result in tree sequences with very few mutations (in the limit, one
    mutation per variable site) and many more recombination events and edges.
     

    .. note::
        As a rough guide,  *relative* mismatch values somewhere between 1e-3 and 1e3
        have been found by simulation to give reasonable results. If absolute mismatch
        likelihoods are specified by using `mismatch_likelihoods` rather than the
        `relative_mismatch` parameter, the appropriate comparison is against the
        recombination rate multiplied by the average distance between inference sites.
        As mismatch likelihoods tend to zero, the inferred tree sequence increases in
        conformity with the infinite sites model of evolution, where a single historical
        mutation event explains which samples have the derived allele at a site.
        This was the default behaviour in versions of tsinfer prior to 0.2.0.
        

    :param SampleData sample_data: The :class:`SampleData` instance
        representing the input data.
    :param AncestorData ancestor_data: The :class:`AncestorData` instance
        representing the set of ancestral haplotypes for which we are finding
        a history.
    :param (float or RateMap) recombination_rate: Either a RateMap object, or a floating
        point value giving a single rate, :math:`rho`, across the entire sequence.
        Integrating this rate between adjacent sites provides a list of relative
        likelihoods of recombination, :math:`r`, which determine the trade off between 
        recombination and multiple mutation during the matching process.
    :param float relative_mismatch: The relative mismatch likelihood for a site. For
        example, if 1 (the default used when the parameter is set to ``None``), then the
        true mismatch likelihood for all sites is set to the median recombination
        likelihood :math:`$\tilde{r}`. For recommended values, see Note above.
    :param bool path_compression: Whether to merge edges that share identical
        paths (essentially taking advantage of shared recombination breakpoints).
    :param int num_threads: The number of match worker threads to use. If
        this is <= 0 then a simpler sequential algorithm is used (default).
    :param int precision: TODO: SOME DESCRIPTION HERE.
    :param (Ratemap or np.array) mismatch_likelihoods: An alternative to
        ``relative_mismatch`` that can be used to set absolute mismatch likelihoods
        directly. If a RateMap object is given, the map value at each site position
        provides the absolute mismatch likelihood. If an array is given, it provides the
        absolute mismatch likelihood for each inference site (and hence must be of the
        length of the number of inference sites, or ``ancestors_data.num_sites``)
    :return: The ancestors tree sequence representing the inferred history
        of the set of ancestors.
    :rtype: tskit.TreeSequence
    """
```

**Comment by jeromekelleher** [comment]
I think it's good, but I'd cut back the documented interface to just ``recombination_rate`` and ``relative_mismatch``. That's sufficient flexibility for what we're doing in the current paper, and that's all we need for now. We don't want to document interfaces that we're not 100% sure of.

I'm not sure about using the ``_likelihood`` suffix on the other variables, as I'm not entirely sure that's what they are. I think it would be better to call them ``recombination`` and ``mismatch`` with the semantics of only accepting a numpy array of the appropriate length (making a mistake of setting ``recombination`` instead of ``recombation_rate`` pretty unlikely. That way our existing code will more-or-less stay working and we have room to open up this more advanced interface in the future if we want it.

**Comment by hyanwong** [comment]
Thanks for looking at this @jeromekelleher - really helpful.

> I think it's good, but I'd cut back the documented interface to just `recombination_rate` and `relative_mismatch`. That's sufficient flexibility for what we're doing in the current paper, and that's all we need for now. We don't want to document interfaces that we're not 100% sure of.

Yes, that's a good point. I'd like to document it "for ourselves only" though, so I guess I'll move that to a comment somehow?

> I'm not sure about using the `_likelihood` suffix on the other variables, as I'm not entirely sure that's what they are. 

Right. I *think* that's what they are (and it's what is written in the explanatory documentation), but it would be as well to check with e.g. Gil. They are equivalent to the transmission and emission probabilities of the HMM aren't they? I appended the "likelihood" to make it completely clear that these aren't rates. But if we aren't documenting them, then it doesn't matter, I think, so happy to remove the suffix.

> with the semantics of only accepting a numpy array of the appropriate length (making a mistake of setting recombination instead of recombation_rate pretty unlikely.

I definitely think we should only accept an array for `recombination`. I think we should, in future, allow the possibility of using a RateMap for the mismatch likelihoods/probabilities, but there's no need to do that now, since we don't use it anyway.

**Comment by hyanwong** [comment]
p.s. I've got the RateMap class over in https://github.com/tskit-dev/msprime/pull/1185 to a state where I'm happy with the interface, although coverage isn't up to the mark yet. After discussion with @grahamgower we left the start position in, and also added an end position, so that we can mark the beginning and end of the documented map (i.e. the rows in the hapmap.txt file) while still having a map that is the complete length of the chromosome. So RateMap now goes:

```
RateMap(self, position, rate=None, cumulative_rate=None, start_position=None, end_position=None)
```

You can either specify `rate` or `cumulative_rate`, but not both - it seemed sensible to allow direct reading from the genetic distance column.

I'll assume this is the way we want to go, and copy & paste the RateMap (and new read_hapmap function) into a separate `intervals.py` file in tsinfer, so I can implement the new recombination & mismatch interface.

**Comment by hyanwong** [comment]
Fixed in #395 , still to be parameterised, though (see #398 )
